### PR TITLE
TICK-192 ProxGroupL1

### PR DESCRIPTION
**Pull request by stephanegaiffas**
This adds a ProxGroupL1 proximal operator and a ProxL2 operator.
This allows to induce group sparsity in the weights of a model.
We need this quickly for the revision of a paper with Simon

### CODE REVIEW

**Comment by stephanegaiffas** [comment]
Ready for review @Mbompr @SimonBussy

File: tick/optim/prox/prox_group_l1.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
+# License: BSD 3 clause
+
+import numpy as np
+from .base import Prox
+from .build.prox import ProxGroupL1 as _ProxGroupL1
+
+
+# TODO: update tests for ProxL2 and ProxGroupL1
+# TODO: check that the prox indeed works with a solver
```
Comment (Mbompr) [inline tick/optim/prox/prox_group_l1.py]: What about these TODO :)?

File: tick/optim/prox/prox_group_l1.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            blocks_length = blocks_length.astype(np.uint64)
+
+        if blocks_start.shape != blocks_length.shape:
+            raise ValueError("``blocks_start`` and ``blocks_length`` "
+                             "must have the same size")
+        if any(blocks_start[1:] < blocks_start[:-1]):
+            raise ValueError('``block_start`` must be sorted')
+
+        if any(blocks_start[1:] < blocks_start[:-1] + blocks_length[:-1]):
+            raise ValueError("blocks must not overlap")
```
Comment (Mbompr) [inline tick/optim/prox/prox_group_l1.py]: could'nt we mutualize this code with prox binarsity? Maybe by creating a stateless function that takes block_length and block_start as input?

Comment (Mbompr) [inline tick/optim/prox/prox_group_l1.py]: Or better, but more time consuming, with inheritance

File: tick/optim/prox/src/prox_group_l1.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
+// License: BSD 3 clause
+
+#include "prox_l2.h"
+#include "prox_group_l1.h"
+
+ProxGroupL1::ProxGroupL1(double strength,
+                             SArrayULongPtr blocks_start,
```
Comment (Mbompr) [inline tick/optim/prox/src/prox_group_l1.cpp]: reformat code

File: tick/optim/prox/src/prox_l2.cpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
+// License: BSD 3 clause
+
+#include "prox_l2.h"
+
+ProxL2::ProxL2(double strength,
```
Comment (Mbompr) [inline tick/optim/prox/src/prox_l2.cpp]: reformat code

File: tick/optim/prox/src/prox_group_l1.h [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  std::vector<ProxL2> proxs;
+
+  // Tells us if the prox is ready (with correctly allocated sub-prox for each blocks).
+  // This is mainly necessary when the user changes the range from python
+  bool is_synchronized;
+
+  void synchronize_proxs();
+
+ public:
+  ProxGroupL1(double strength, SArrayULongPtr blocks_start, SArrayULongPtr blocks_length,
```
Comment (Mbompr) [inline tick/optim/prox/src/prox_group_l1.h]: reformat code

**Comment by stephanegaiffas** [comment]
Ready @Mbompr, used the inheritance approach (both for C++ and python)

File: tick/optim/prox/base/prox_with_groups.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+class ProxWithGroups(Prox):
+    """Base class of a proximal operator with groups. It applies specific
+    proximal operator in each group, or block. Blocks (non-overlapping) are
+    specified by the ``blocks_start`` and ``blocks_length`` parameters.
+    This base class is not intented for end-users, but for developers only.
+
+    Parameters
+    ----------
+    strength : `float`, default=0.
```
Comment (Mbompr) [inline tick/optim/prox/base/prox_with_groups.py]: strength has no default

File: tick/optim/prox/src/prox_binarsity.cpp [hunk 7]
```diff
@@ -7,76 +7,35 @@ ProxBinarsity::ProxBinarsity(double strength,
...
-    ulong start = (*blocks_start)[k];
-    if (has_range) {
-      // If there is a range, we apply the global start
-      start += this->start;
-    }
-    ulong end = start + (*blocks_length)[k];
-    blocks_prox.push_back(ProxTV(strength, start, end, positive));
-  }
-  prox_tv_synchronized = true;
+std::unique_ptr<Prox> ProxBinarsity::new_prox(double strength, ulong start, ulong end, bool positive) {
```
Comment (Mbompr) [inline tick/optim/prox/src/prox_binarsity.cpp]: I would rather call it build_prox

File: tick/optim/prox/src/prox_with_groups.cpp [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+void ProxWithGroups::synchronize_proxs() {
+  proxs.clear();
+  for (ulong k = 0; k < n_blocks; k++) {
+    ulong start = (*blocks_start)[k];
+    if (has_range) {
+      // If there is a range, we apply the global start
+      start += this->start;
+    }
+    ulong end = start + (*blocks_length)[k];
+    proxs.emplace_back(new_prox(strength, start, end, positive));
```
Comment (Mbompr) [inline tick/optim/prox/src/prox_with_groups.cpp]: Nice one :)!

File: tick/optim/prox/src/prox_group_l1.h [hunk 9]
```diff
@@ -0,0 +1,22 @@
+// License: BSD 3 clause
```
Comment (Mbompr) [inline tick/optim/prox/src/prox_group_l1.h]: License should be after header guards

File: tick/optim/prox/src/prox_with_groups.h [hunk 10]
```diff
@@ -0,0 +1,78 @@
+// License: BSD 3 clause
```
Comment (Mbompr) [inline tick/optim/prox/src/prox_with_groups.h]: License should be after header guards

**Pull request merged**